### ISSUE TITLE
Validation data in GridSearch

### ISSUE THREAD

**Issue by bright1993ff66** [skorch-dev/skorch#622]
Hi, I am quite new to Skorch and I am tuning hyperparameters using GridSearch. I have created my own training data, validation data and test data based on ```torch.utils.data.Dataset``` class. 

```Python
class Data(th.utils.data.Dataset):
    def __init__(self, X, y):
        self.X = X
        self.Y = y
        
    def __getitem__(self, idx):
        return self.X[idx], self.Y[idx]
    
    def __len__(self):
        return self.X.shape[0]

train_data = Data(X = train_x_tensor, y=train_y_tensor)
valid_data = Data(X = valid_x_tensor, y=valid_y_tensor)
test_data = Data(X = test_x_tensor, y=test_y_tensor)
```
Then I use the ```NeuralNet``` class to create net for training and validation:

```Python
early_stopping_callbacks = EarlyStopping(monitor='valid_loss', lower_is_better=True, patience=5)

net = NeuralNet(
    ClassifierModule,
    max_epochs=20,
    optimizer=th.optim.Adam,
    lr=0.01,
    device=device,  # uncomment this to train with CUDA
    criterion=nn.BCELoss, 
    train_split=predefined_split(valid_data), 
    callbacks=[early_stopping_callbacks],
)
```

And I specify the following hyperparameters to tune:

```Pyhon
params = {
    'lr': [0.05, 0.1],
    'module__num_units': [10, 20],
    'module__dropout': [0, 0.5],
    'optimizer__weight_decay': [0, 1e-04],
    'iterator_train__batch_size': [32, 64],
    'iterator_valid__batch_size': [valid_data.X.shape[0]]
}
```

**My question is**, when I conducting the **GridSearch** like the following:

```Python
gs = GridSearchCV(net, params, refit=False, cv=3, scoring='roc_auc', verbose=2)
gs.fit(train_data)
```

1. Does the ```GridSearchCV``` use the valid_data specified in here: ```train_split=predefined_split(valid_data)``` to conduct validation?
2. When doing the GridSearch, for each hyperparameter setting, would the model still stop when early stopping is triggered?
3. When I want to know how my model performs on the whole validation data, is it correct to specify the size of validation data in the ```params``` like ```'iterator_valid__batch_size': [valid_data.X.shape[0]]```?

Thank you very much!

**Comment by ottonemo** [comment]
Hi,

just a quick note regarding your `Data` object: there is [torch.utils.data.TensorDataset](https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#TensorDataset) which does essentially the same (except of course this was for demonstration purposes, then ignore my comment).

To answer your questions quickly:

> Does the `GridSearchCV` use the `valid_data` specified in here: `train_split=predefined_split(valid_data)` to conduct validation?

No, `GridSearchCV` will take whatever you supply in `gs.fit()` and will apply the cross-validator that you supply when calling `GridSearchCV()`.

> When doing the GridSearch, for each hyperparameter setting, would the model still stop when early stopping is triggered?

Yes, as long as there is a validation set that skorch can use to compute validation scores the early stopping callback will work.

> When I want to know how my model performs on the whole validation data, is it correct to specify the size of validation data in the params like 'iterator_valid__batch_size': [valid_data.X.shape[0]]?

It depends on the circumstances but generally I would say: no. Scores computed via `EpochScoring` callbacks take the whole data set into account anyway (so there are no batching artifacts like in `BatchScoring`). Setting the batch size to the size of the whole data set would just hinder performance for other tasks such as inference via `predict`.

More elaborate explanation:

skorch has an internal train/validation split to support things like early stopping which is a bit cumbersome to implement in sklearn. As a consequence there is no back-channel to communicate any validation sets to objects like `GridSearchCV` but that doesn't matter since you wouldn't want to do a grid search with a fixed train/validation split anyway (you'd want a k-fold split in most cases anyway).

What you can do is to let skorch split a portion from the input training data on its own (e.g. by supplying a cv split with the `train_split` parameter of `skorch.NeuralNet`). This way you can choose an arbitrary cross validation strategy for your grid search (e.g. `StratifiedKFold`, `TimeSeriesSplit`, ...) and whatever gets supplied to `net.fit()` is then split again into training and validation data. Of course, this has the drawback of having less data to train on but you gain early stopping.

**Comment by bright1993ff66** [comment]
Thank you very much for your detailed explanation! You have answered all my questions. I will close this issue~

> Hi,
> 
> just a quick note regarding your `Data` object: there is [torch.utils.data.TensorDataset](https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#TensorDataset) which does essentially the same (except of course this was for demonstration purposes, then ignore my comment).
> 
> To answer your questions quickly:
> 
> > Does the `GridSearchCV` use the `valid_data` specified in here: `train_split=predefined_split(valid_data)` to conduct validation?
> 
> No, `GridSearchCV` will take whatever you supply in `gs.fit()` and will apply the cross-validator that you supply when calling `GridSearchCV()`.
> 
> > When doing the GridSearch, for each hyperparameter setting, would the model still stop when early stopping is triggered?
> 
> Yes, as long as there is a validation set that skorch can use to compute validation scores the early stopping callback will work.
> 
> > When I want to know how my model performs on the whole validation data, is it correct to specify the size of validation data in the params like 'iterator_valid__batch_size': [valid_data.X.shape[0]]?
> 
> It depends on the circumstances but generally I would say: no. Scores computed via `EpochScoring` callbacks take the whole data set into account anyway (so there are no batching artifacts like in `BatchScoring`). Setting the batch size to the size of the whole data set would just hinder performance for other tasks such as inference via `predict`.
> 
> More elaborate explanation:
> 
> skorch has an internal train/validation split to support things like early stopping which is a bit cumbersome to implement in sklearn. As a consequence there is no back-channel to communicate any validation sets to objects like `GridSearchCV` but that doesn't matter since you wouldn't want to do a grid search with a fixed train/validation split anyway (you'd want a k-fold split in most cases anyway).
> 
> What you can do is to let skorch split a portion from the input training data on its own (e.g. by supplying a cv split with the `train_split` parameter of `skorch.NeuralNet`). This way you can choose an arbitrary cross validation strategy for your grid search (e.g. `StratifiedKFold`, `TimeSeriesSplit`, ...) and whatever gets supplied to `net.fit()` is then split again into training and validation data. Of course, this has the drawback of having less data to train on but you gain early stopping.
